### PR TITLE
KDFs: use generated parameter parsing

### PR DESCRIPTION
**Pull request by paulidale**
KDFs are OSSL_PARAM heavy and widely used.  Switching to a generated name parser should improve performance a bit.  We'll also gain forced consistency between advertised parameter lists and what is actually accepted.

This requires some changes to the parser generator to support duplicated parameters.
It also requires changes to the KDF code to use the parser.
If additionally requires changes to some of the helper functions so they don't attempt to locate parameters.

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
Should be ready for reviewing.

**Comment by paulidale** [comment]
Rebased to resolve trivial conflict.  Reviews @openssl/committers ?

File: providers/implementations/kdfs/hkdf.c.in [hunk 1]
```diff
@@ -306,46 +300,60 @@ static int hkdf_common_set_ctx_params(KDF_HKDF *ctx, const OSSL_PARAM params[])
...
+{- produce_param_decoder('hkdf_set_ctx_params',
+                         (['KDF_PARAM_MODE',           'mode',   'utf8_string'],
+                          ['KDF_PARAM_MODE',           'mode',   'int'],
+                          ['KDF_PARAM_PROPERTIES',     'propq',  'utf8_string'],
+                          ['ALG_PARAM_ENGINE',         'engine', 'utf8_string', 'hidden'],
+                          ['KDF_PARAM_DIGEST',         'digest', 'utf8_string'],
+                          ['KDF_PARAM_KEY',            'key',    'octet_string'],
+                          ['KDF_PARAM_SALT',           'salt',   'octet_string'],
+                          ['KDF_PARAM_FIPS_KEY_CHECK', 'ind_k',  'int'],
+                          ['KDF_PARAM_INFO',           'info',   'octet_string', 5],
```
Comment (t8m) [inline providers/implementations/kdfs/hkdf.c.in]: This is all great piece of work. I am concerned about just one thing. Isn't it too limiting to have a constant maximum number of concatenated params of the same name? I mean the performance for the concatenated usage perhaps isn't too relevant. Could we just return the first or the first two params for the concatenation and require using locate to retrieve any subsequent ones? Or would it be too hard to implement with the current param decoder?

Comment (paulidale) [reply providers/implementations/kdfs/hkdf.c.in]: The main use for this is for TLS where five is the magic number.  For all except TLS1PRF, the only place where multiples of one parameter are given is in our CIs, so it's entirely possible that this isn't an issue at all.  I don't know why duplicates are permitted so widely, possibly copy-paste or misinterpreting things.

It would be possible to do something but the point of this conversion is to avoid processing param lists more than once.  I also worry about complicating things for something that's very unlikely to occur.

A compromise might be to make the arrays larger.  Say 10 or 20 or 50 (or ??) which should be enough to cover reasonable practical uses.  The cost will be zeroing that much more memory each time.  There is no possibility of silent failure in the event of too many repeats, the generated code returns an error if the array would overflow.

**Comment by paulidale** [comment]
Rebase and some fixes because the fixed digest HKDF PR #27247 was merged.

File: providers/implementations/ciphers/cipher_chacha20_poly1305.c.in [hunk 2]
```diff
@@ -187,10 +188,8 @@ static int chacha20_poly1305_set_ctx_params(void *vctx,
         (PROV_CIPHER_HW_CHACHA20_POLY1305 *)ctx->base.hw;
     struct chacha20_poly1305_set_ctx_params_st p;
 
-    if (ossl_param_is_empty(params))
-        return 1;
-
-    p = chacha20_poly1305_set_ctx_params_decoder(params);
+    if (ctx == NULL || !chacha20_poly1305_set_ctx_params_decoder(params, &p))
```
Comment (slontis) [inline providers/implementations/ciphers/cipher_chacha20_poly1305.c.in:191]: ossl_param_is_empty() caused an early out return before. Should it remain?

File: util/perl/OpenSSL/paramnames.pm [hunk 3]
```diff
@@ -644,28 +644,49 @@ sub generate_public_macros {
...
         if (not $case_sensitive) {
             $suf =~ tr/_/-/;
             print " || $strcmp(\"$suf\", s + $n) == 0"
                 if ($suf ne $trieref->{'suffix'});
         }
-        printf "))\n%sr.%s = (OSSL_PARAM *)p;\n",
-               $indent1, $field;
+        print ")) {\n";
+        trie_matched($field, $num, $indent1, $indent2);
+       printf "%s}\n", $indent0;
```
Comment (slontis) [inline util/perl/OpenSSL/paramnames.pm]: bad indent here?

File: providers/implementations/ciphers/cipher_chacha20_poly1305.c.in [hunk 4]
```diff
@@ -113,7 +113,8 @@ static int chacha20_poly1305_get_ctx_params(void *vctx, OSSL_PARAM params[])
     PROV_CHACHA20_POLY1305_CTX *ctx = (PROV_CHACHA20_POLY1305_CTX *)vctx;
     struct chacha20_poly1305_get_ctx_params_st p;
 
-    p = chacha20_poly1305_get_ctx_params_decoder(params);
+    if (ctx == NULL || !chacha20_poly1305_get_ctx_params_decoder(params, &p))
```
Comment (slontis) [inline providers/implementations/ciphers/cipher_chacha20_poly1305.c.in:116]: ctx NULL checking seems unnecesary?

**Review by slontis — COMMENTED** [review]
The code that the decoder produces is of the form:
````
                case 'i':
                    if (ossl_likely(strcmp("pher", s + 2) == 0)) {
                        if (ossl_unlikely(r->cipher != NULL))
                            return 0;
                        r->cipher = (OSSL_PARAM *)p;
                    }
````
Where it returns 0 here if the parameter is repeated, seems like a change in behaviour from what it did before??
Is this change potentially going to break someone?

**Comment by slontis** [comment]
What if for example the user was trying to set something back to NULL?

File: providers/implementations/ciphers/cipher_chacha20_poly1305.c.in (hunk 2 above)
Comment (paulidale) [reply providers/implementations/ciphers/cipher_chacha20_poly1305.c.in:191]: It's not required any more.  Empty params clear all the fields in the structure and the subsequent `!= NULL` checks will all failed and nothing will be processed.  The cost is the series of `!= NULL` which isn't much in what should be a rare situation.  I figured this wasn't enough to justify having the additional call to check for empty and it does neaten up the code in a lot of places.

File: util/perl/OpenSSL/paramnames.pm (hunk 3 above)
Comment (paulidale) [reply util/perl/OpenSSL/paramnames.pm]: Yes :(

File: providers/implementations/ciphers/cipher_chacha20_poly1305.c.in (hunk 4 above)
Comment (paulidale) [reply providers/implementations/ciphers/cipher_chacha20_poly1305.c.in:116]: It was deemed desirable to consistently check for this in an earlier PR I did.  Thus, I've added it throughout.  What was there before was a real hodge podge.

**Comment by paulidale** [comment]
> Where it returns 0 here if the parameter is repeated, seems like a change in behaviour from what it did before?? Is this change potentially going to break someone?

The change is from ignoring the later dupicated parameters to returning an error when they are present.
Currently if a user supplies the same parameter multiple times, only the first will ever be located and used.  Later copies will be ignored.  IMO, this is a user error and we're better off reporting that than silently ignoring it.

Note: the places where multiple copies of a single parameter are allowed are processed differently and the duplicates are dealt with correctly.

**Comment by beldmit** [comment]
> > Where it returns 0 here if the parameter is repeated, seems like a change in behaviour from what it did before?? Is this change potentially going to break someone?
> 
> The change is from ignoring the later dupicated parameters to returning an error when they are present. Currently if a user supplies the same parameter multiple times, only the first will ever be located and used. Later copies will be ignored. IMO, this is a user error and we're better off reporting that than silently ignoring it.
> 
> Note: the places where multiple copies of a single parameter are allowed are processed differently and the duplicates are dealt with correctly.

It definitely looks like a potentially breaking change to me, especially in complex environments.

**Comment by paulidale** [comment]
Is catching and reporting user error a breaking change?  Changing it so it behaves the same as before is easy.  I just don't think it's a good idea to let mistakes propagate.

**Comment by beldmit** [comment]
We may need some "upsert" parameters API for replacement of the value in the existing array then. 

It also makes sense to not throw an error when the old value matches the original one

**Comment by paulidale** [comment]
I don't understand.  Could you give an example?

**Comment by beldmit** [comment]
For upsert I considered the scenario when we append parameters to an array on and on, but it is not relevant.

If we set the parameters one by one, none of my ideas would work

**Comment by paulidale** [comment]
Appending to an array and setting parameters multiple times wouldn't have worked in the past.  Now it would generate an error when the duplicate appears.  Appending and setting once would now error rather than ignoring the duplicates.

Setting one by one will still work fine.

Nonetheless, I've just pushed an update that returns to the former (broken) behaviour.  I don't like suppressing errors silently but it will be 100% compatible.

**Comment by paulidale** [comment]
> What if for example the user was trying to set something back to NULL?

There would be no difference.  The parameter would have a non-null address and this is what the decoder would set.
The actual setting to null isn't part of the decoder's duties.

**Comment by paulidale** [comment]
I'll raise the error on duplication as a separate PR after this one merges.  We can discuss it then.

File: .gitignore [hunk 5]
```diff
@@ -84,6 +84,19 @@ providers/common/include/prov/der_digests.h
...
+providers/implementations/kdfs/kbkdf.c
+providers/implementations/kdfs/krb5kdf.c
+providers/implementations/kdfs/pbkdf1.c
+providers/implementations/kdfs/pbkdf2.c
+providers/implementations/kdfs/pkcs12kdf.c
+providers/implementations/kdfs/pvkkdf.c
+providers/implementations/kdfs/scrypt.c
+providers/implementations/kdfs/sshkdf.c
+providers/implementations/kdfs/sskdf.c
+providers/implementations/kdfs/tls1_prf.c
```
Comment (slontis) [inline .gitignore:99]: Is providers/implementations/kdfs/x942kdf.c missing?

File: providers/common/provider_util.c [hunk 6]
```diff
@@ -48,17 +48,14 @@ int ossl_prov_cipher_copy(PROV_CIPHER *dst, const PROV_CIPHER *src)
     return 1;
 }
 
-static int load_common(const OSSL_PARAM params[], const char **propquery,
-                       ENGINE **engine)
+static int load_common(const OSSL_PARAM *propq, const OSSL_PARAM *e,
```
Comment (slontis) [inline providers/common/provider_util.c]: Not sure how "common" this function is now. This is handling specific parameters now. Maybe these are just separate functions now?

File: providers/common/provider_util.c [hunk 7]
```diff
@@ -88,33 +84,29 @@ static int load_common(const OSSL_PARAM params[], const char **propquery,
     return 1;
 }
 
-int ossl_prov_cipher_load_from_params(PROV_CIPHER *pc,
-                                      const OSSL_PARAM params[],
-                                      OSSL_LIB_CTX *ctx)
+int ossl_prov_cipher_load(PROV_CIPHER *pc, const OSSL_PARAM *c,
```
Comment (slontis) [inline providers/common/provider_util.c]: c is a horrible name :)

File: providers/common/provider_util.c [hunk 8]
```diff
@@ -211,6 +212,17 @@ int ossl_prov_digest_load_from_params(PROV_DIGEST *pd,
     return pd->md != NULL;
 }
 
+int ossl_prov_digest_load_from_params(PROV_DIGEST *pd,
+                                      const OSSL_PARAM params[],
+                                      OSSL_LIB_CTX *ctx)
+{
+    return ossl_prov_digest_load(pd,
```
Comment (slontis) [inline providers/common/provider_util.c:219]: Why does the cipher call do the is_empty check and this one doesnt?

File: .gitignore (hunk 5 above)
Comment (paulidale) [reply .gitignore:99]: Yes :disappointed:

File: providers/common/provider_util.c (hunk 8 above)
Comment (paulidale) [reply providers/common/provider_util.c:219]: An oops.  Removed the check in the cipher call.

File: providers/common/provider_util.c (hunk 6 above)
Comment (paulidale) [reply providers/common/provider_util.c]: Split and renamed.

File: providers/common/provider_util.c [hunk 9]
```diff
@@ -335,7 +340,16 @@ int ossl_prov_macctx_load_from_params(EVP_MAC_CTX **macctx,
...
+    if (ciphername == NULL && pcipher != NULL)
+        if (!OSSL_PARAM_get_utf8_string_ptr(pcipher, &ciphername))
+            return 0;
+    if (mdname == NULL && pdigest != NULL)
+        if (!OSSL_PARAM_get_utf8_string_ptr(pdigest, &mdname))
+            return 0;
+    if (pengine != NULL && !OSSL_PARAM_get_utf8_string_ptr(pengine, &engine))
+        return 0;
+
+    if (ossl_prov_set_macctx(*macctx, NULL, ciphername, mdname, engine,
```
Comment (slontis) [inline providers/common/provider_util.c]: It looks like this function is only called from one other place, and that also passes NULL as the params - so it looks like that is redundant and can be removed from the function.

File: providers/implementations/ciphers/cipher_chacha20_poly1305.c.in (hunk 4 above)
Comment (slontis) [reply providers/implementations/ciphers/cipher_chacha20_poly1305.c.in:116]: Looks like EVP_CipherInit_ex2() sets ctx->algctx to NON NULL or it fails.
It would be NULL after EVP_CIPHER_CTX_new() or EVP_CIPHER_CTX_reset(). What is the case this check is needed for?

File: providers/common/provider_util.c (hunk 9 above)
Comment (paulidale) [reply providers/common/provider_util.c]: This is what was there but I've updated the function removing "unused" arguments.

File: providers/implementations/ciphers/cipher_chacha20_poly1305.c.in (hunk 4 above)
Comment (paulidale) [reply providers/implementations/ciphers/cipher_chacha20_poly1305.c.in:116]: Best ask those who insisted on adding these checks in the earlier PR for the case they thought important.

This is a public API where checking arguments is good practice.  

~It only took a few seconds to uncover one case where it is useful: `EVP_CIPHER_CTX_get_params(EVP_CIPHER_CTX_new(...), ...)`.  I'd be surprised if there weren't other paths to a similar result.~ (don't think too much after bedtime)

File: providers/implementations/kdfs/hkdf.c.in [hunk 10]
```diff
@@ -385,111 +389,96 @@ static int kdf_hkdf_set_ctx_params(void *vctx, const OSSL_PARAM params[])
...
-    };
-    return known_settable_ctx_params;
+    return hkdf_set_ctx_params_list;
 }
 
-static int hkdf_common_get_ctx_params(KDF_HKDF *ctx, OSSL_PARAM params[])
+{- produce_param_decoder('hkdf_get_ctx_params',
+                         (['KDF_PARAM_SIZE',                    'size',   'size_t'],
+                          ['KDF_PARAM_DIGEST',                  'digest', 'utf8_string'],
+                          ['KDF_PARAM_MODE',                    'mode',   'utf8_string'],
```
Comment (slontis) [inline providers/implementations/kdfs/hkdf.c.in:398]: Does the generated output for this multiple use of 'mode' work?

Comment (paulidale) [inline providers/implementations/kdfs/hkdf.c.in:399]: They collapse into one path in the TRIE, so either flavour is found.  It's up to the calling code to deal with the different types.

File: providers/implementations/kdfs/hkdf.c.in [hunk 11]
```diff
@@ -335,46 +325,60 @@ static int hkdf_common_set_ctx_params(KDF_HKDF *ctx, const OSSL_PARAM params[])
...
+
+{- produce_param_decoder('hkdf_set_ctx_params',
+                         (['KDF_PARAM_MODE',           'mode',   'utf8_string'],
+                          ['KDF_PARAM_MODE',           'mode',   'int'],
+                          ['KDF_PARAM_PROPERTIES',     'propq',  'utf8_string'],
+                          ['ALG_PARAM_ENGINE',         'engine', 'utf8_string', 'hidden'],
+                          ['KDF_PARAM_DIGEST',         'digest', 'utf8_string'],
+                          ['KDF_PARAM_KEY',            'key',    'octet_string'],
+                          ['KDF_PARAM_SALT',           'salt',   'octet_string'],
+                          ['KDF_PARAM_FIPS_KEY_CHECK', 'ind_k',  'int'],
```
Comment (slontis) [inline providers/implementations/kdfs/hkdf.c.in:362]: The generated code for key-check seems to be a bit funky for some reason:

````
            case 'k':
                switch(s[1]) {
                default:
                    break;
                case 'e':
                    switch(s[2]) {
                    default:
                        break;
                    case 'y':
                        switch(s[3]) {
                        default:
                            break;
                        case '-':
                            if (ossl_likely(strcmp("check", s + 4) == 0)) {
                                if (ossl_unlikely(r->ind_k != NULL))
                                    return 0;
                                r->ind_k = (OSSL_PARAM *)p;
                            }
                            break;
                        case '\0':
                            if (ossl_unlikely(r->key != NULL))
                                return 0;
                            r->key = (OSSL_PARAM *)p;
                        }
                    }
                }
                break;
````

Not sure why its different than the others which are of the form

strcmp("ey-check", s+1) == 0)

Unless its related to it containing a "-"

File: providers/implementations/kdfs/hkdf.c.in (hunk 1 above)
Comment (slontis) [reply providers/implementations/kdfs/hkdf.c.in]: We should probably at least spit out an error in the event the buffer is full rather than just return 0

Comment (paulidale) [reply providers/implementations/kdfs/hkdf.c.in]: Added an error in addition to the zero return.

File: providers/implementations/kdfs/hkdf.c.in [hunk 12]
```diff
@@ -966,61 +1003,77 @@ static int kdf_tls1_3_derive(void *vctx, unsigned char *key, size_t keylen,
...
+
+{- produce_param_decoder('kdf_tls1_3_set_ctx_params',
+                         (['KDF_PARAM_MODE',              'mode',   'utf8_string'],
+                          ['KDF_PARAM_MODE',              'mode',   'int'],
+                          ['KDF_PARAM_PROPERTIES',        'propq',  'utf8_string'],
+                          ['ALG_PARAM_ENGINE',            'engine', 'utf8_string', 'hidden'],
+                          ['KDF_PARAM_DIGEST',            'digest', 'utf8_string'],
+                          ['KDF_PARAM_KEY',               'key',    'octet_string'],
+                          ['KDF_PARAM_SALT',              'salt',   'octet_string'],
+                          ['KDF_PARAM_FIPS_KEY_CHECK',    'ind_k',  'int'],
```
Comment (slontis) [inline providers/implementations/kdfs/hkdf.c.in:1017]: I would prefer if the fips indicators were not abbreviated..
since we have to map SETTABLE indexes to these values..
the name should be something that maps nicely to FIPS_KEY_CHECK..

We have bits of code that do this....

````
        if (!OSSL_FIPS_IND_ON_UNAPPROVED(ctx, OSSL_FIPS_IND_SETTABLE0,
                                         libctx, "HKDF", "Key size",
                                         ossl_fips_config_hkdf_key_check)) {
            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
````
and
````
    if (!OSSL_FIPS_IND_SET_CTX_PARAM(ctx, OSSL_FIPS_IND_SETTABLE0, params,
                                     OSSL_KDF_PARAM_FIPS_KEY_CHECK))
````

File: providers/implementations/kdfs/hkdf.c.in [hunk 13]
```diff
@@ -281,25 +263,33 @@ static int kdf_hkdf_derive(void *vctx, unsigned char *key, size_t keylen,
...
+    OSSL_PARAM *engine;
+    OSSL_PARAM *digest;
+    OSSL_PARAM *key;
+    OSSL_PARAM *salt;
+    OSSL_PARAM *ind_k;
+    OSSL_PARAM *ind_d;
+    OSSL_PARAM *prefix;
+    OSSL_PARAM *label;
+    OSSL_PARAM *data;
+    OSSL_PARAM *info[5];
```
Comment (slontis) [inline providers/implementations/kdfs/hkdf.c.in]: At least #define this magic number and use it instead of 5 in multiple places?

File: providers/implementations/kdfs/hkdf.c.in (hunk 12 above)
Comment (paulidale) [reply providers/implementations/kdfs/hkdf.c.in:1017]: I don't agree.  To me `ind_k` and `ind_d` are good enough for key and data respectively.  Where there is only one such indicator, I'd consider changing the name to `ind` since it's unambiguous.

The only linking commonality between the two fragments presented is `OSSL_FIPS_IND_SETTABLE0` which bears no relation with what the indicator is actually doing or is used for.  IMO we should address this disconnect before worrying about structure field names.  However, it would make sense to change the field names to `ind0` & `ind1` to align with the common `OSSL_FIPS_IND_SETTABLE` macros.

Comment (slontis) [reply providers/implementations/kdfs/hkdf.c.in:1017]: That is exactly the point. We are mapping a index to a name, and the name is in multiple places.

File: providers/implementations/kdfs/hkdf.c.in [hunk 14]
```diff
@@ -966,61 +1004,77 @@ static int kdf_tls1_3_derive(void *vctx, unsigned char *key, size_t keylen,
...
     }
 
-    OPENSSL_clear_free(ctx->data, ctx->data_len);
-    ctx->data = NULL;
-    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_DATA)) != NULL
-            && !OSSL_PARAM_get_octet_string(p, (void **)&ctx->data, 0,
-                                            &ctx->data_len))
-        return 0;
+    if (p.data != NULL) {
+        OPENSSL_clear_free(ctx->data, ctx->data_len);
```
Comment (slontis) [inline providers/implementations/kdfs/hkdf.c.in:1062]: Hmm this was always clearing data before which seems wrong. Unless it was deliberate..

File: providers/implementations/kdfs/kbkdf.c.in [hunk 15]
```diff
@@ -363,22 +366,41 @@ static int kbkdf_derive(void *vctx, unsigned char *key, size_t keylen,
     return ret;
 }
 
+{- produce_param_decoder('kbkdf_set_ctx_params',
+                         (['KDF_PARAM_INFO',                'info',   'octet_string', 5],
```
Comment (slontis) [inline providers/implementations/kdfs/kbkdf.c.in]: Magic number 5 again..

File: providers/implementations/kdfs/kbkdf.c.in [hunk 16]
```diff
@@ -363,22 +366,41 @@ static int kbkdf_derive(void *vctx, unsigned char *key, size_t keylen,
...
+                          ['KDF_PARAM_DIGEST',              'digest', 'utf8_string'],
+                          ['KDF_PARAM_CIPHER',              'cipher', 'utf8_string'],
+                          ['KDF_PARAM_MAC',                 'mac',    'utf8_string'],
+                          ['KDF_PARAM_MODE',                'mode',   'utf8_string'],
+                          ['KDF_PARAM_PROPERTIES',          'propq',  'utf8_string'],
+                          ['ALG_PARAM_ENGINE',              'engine', 'utf8_string', 'hidden'],
+                          ['KDF_PARAM_KBKDF_USE_L',         'use_l',  'int'],
+                          ['KDF_PARAM_KBKDF_USE_SEPARATOR', 'sep',    'int'],
+                          ['KDF_PARAM_KBKDF_R',             'r',      'int'],
+                          ['KDF_PARAM_FIPS_KEY_CHECK',      'ind_d',  'int'],
```
Comment (slontis) [inline providers/implementations/kdfs/kbkdf.c.in]: 'ind_d' is definitely not the right thing.

File: providers/implementations/kdfs/kbkdf.c.in (hunk 15 above)
Comment (paulidale) [reply providers/implementations/kdfs/kbkdf.c.in]: There is only one instance of this in the file so I figured it didn't need to be a macro.  Easy to change to one.

File: providers/implementations/kdfs/hkdf.c.in (hunk 14 above)
Comment (paulidale) [reply providers/implementations/kdfs/hkdf.c.in:1062]: It's cleared in the free too.  I don't enough about the usage of this parameter to know if it needs to be cleared or not.  Suggest that would be a different PR if we want to change it.

File: providers/implementations/kdfs/kbkdf.c.in [hunk 17]
```diff
@@ -462,51 +480,35 @@ static int kbkdf_set_ctx_params(void *vctx, const OSSL_PARAM params[])
...
-        OSSL_FIPS_IND_SETTABLE_CTX_PARAM(OSSL_KDF_PARAM_FIPS_KEY_CHECK)
-        OSSL_PARAM_END,
-    };
-    return known_settable_ctx_params;
+    return kbkdf_set_ctx_params_list;
 }
 
+{- produce_param_decoder('kbkdf_get_ctx_params',
+                         (['KDF_PARAM_SIZE',                    'size', 'size_t'],
+                          ['ALG_PARAM_FIPS_APPROVED_INDICATOR', 'ind',  'int'],
```
Comment (slontis) [inline providers/implementations/kdfs/kbkdf.c.in]: Maybe keep the KDF alias for the indicator here..

File: providers/implementations/kdfs/krb5kdf.c.in [hunk 18]
```diff
@@ -151,69 +154,73 @@ static int krb5kdf_derive(void *vctx, unsigned char *key, size_t keylen,
...
+
+        if (cipher != NULL)
+            len = EVP_CIPHER_get_key_length(cipher);
+        else
+            len = EVP_MAX_KEY_LENGTH;
+
+        if (!OSSL_PARAM_set_size_t(p.size, len))
+            return 0;
+    }
+    return 1;
```
Comment (slontis) [inline providers/implementations/kdfs/krb5kdf.c.in:217]: This return is a change in behaviour - but looks like a bug that I thought I had already fixed previously.

Comment (paulidale) [reply providers/implementations/kdfs/krb5kdf.c.in:217]: We've not been very consistent about this.  Since unknowned parameter should be ignored, returning -2 is never correct.

File: providers/implementations/kdfs/kbkdf.c.in (hunk 17 above)
Comment (paulidale) [reply providers/implementations/kdfs/kbkdf.c.in]: I detest the aliases.  We never should have used them :(

File: providers/implementations/kdfs/hkdf.c.in (hunk 12 above)
Comment (paulidale) [reply providers/implementations/kdfs/hkdf.c.in:1017]: We should fix the index nomenclature and avoid the problem completely.  Have a block of defines at the top of the file like this:
```c
#define HKDF_FIPS_IND_KEY_CHECK OSSL_FIPS_IND_SETTABLE0
```
Adding a kludge that duplicates things & has to be kept synchronised is not the right solution.

**Comment by slontis** [comment]
Look good to me..

**Comment by paulidale** [comment]
Rebased to master and squashed down.

**Review by slontis — APPROVED** [review]

**Comment by paulidale** [comment]
Second review @openssl/committers ?

**Comment by paulidale** [comment]
Second ping for second review please?

**Comment by paulidale** [comment]
Third ping for second review please?

**Comment by paulidale** [comment]
Fourth ping for reviews from the @openssl/committers 

Please.  There are a lot of changes here but they are fairly mechanical in nature.

I've got a pile more similar changes queued up after (and depending on) this.  Currently in #28047 but will likely be split for easier reviewing.

This work fixes a plethora of mismatches and bugs in the various algorithm parameter lists & processing.  Circa half of all algorithms updated have had problems in this area, so the fixes are widespread.  Moreover, due to the use of generated decoders and parameter lists, similar bugs should be avoided moving forwards.

There will also be a performance improvement by avoiding repeated `strcmp(3)` calls to decode parameters.  This improvement has not been quantified.

File: test/evp_kdf_test.c [hunk 19]
```diff
@@ -295,12 +295,18 @@ static int do_kdf_hkdf_gettables(int extract_only, int has_digest)
...
+
     params_get[0] = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_MODE, mode_utf8,
                                                      sizeof(mode_utf8));
-    params_get[1] = OSSL_PARAM_construct_int(OSSL_KDF_PARAM_MODE, &mode_int);
-    params_get[2] = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SALT, salt, sizeof(salt));
-    params_get[3] = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_INFO, info, sizeof(info));
-    params_get[4] = OSSL_PARAM_construct_end();
+    params_get[1] = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SALT, salt, sizeof(salt));
+    params_get[2] = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_INFO, info, sizeof(info));
+    params_get[3] = OSSL_PARAM_construct_end();
```
Comment (t8m) [inline test/evp_kdf_test.c:309]: Hmm... but this looks like an API break. I know it is a very bad corner case but...

File: test/evp_kdf_test.c [hunk 20]
```diff
@@ -2066,9 +2072,14 @@ static int test_kdf_hmac_drbg_settables(void)
     /* Fail if we pass the wrong type for params */
     params[1] = OSSL_PARAM_construct_end();
     for (i = 0; settableparams[i].key != NULL; ++i) {
-        /* Skip "properties" key since it returns 1 unless the digest is also set */
+        /*
+         * Skip "properties" and "engine" keys since they returns 1 unless
```
Comment (t8m) [inline test/evp_kdf_test.c:2076]: Nit: returns -> return

File: test/evp_kdf_test.c (hunk 19 above)
Comment (paulidale) [reply test/evp_kdf_test.c:309]: This isn't an API break but it is kind of subtle.  It's actually fixing a bug in the test suite.  The difference is that an error is raised for a duplicated parameter whereas perviously the second was silently ignored.

Think about it this way: the receiving code that allows two different "mode" types does an `OSSL_PARAM_locate()` call to find the parameter and this only ever finds the first.  The second is never looked for and is completely ignored.

File: test/evp_kdf_test.c (hunk 20 above)
Comment (paulidale) [reply test/evp_kdf_test.c:2076]: Fixed.

File: test/evp_kdf_test.c (hunk 19 above)
Comment (t8m) [reply test/evp_kdf_test.c:309]: Hmm... you seem to be right, but how could the checks on lines 305 and 306 work then????

Comment (paulidale) [reply test/evp_kdf_test.c:309]: Careful choice of test cases?

Comment (t8m) [reply test/evp_kdf_test.c:309]: But we have test cases with both variants of `extract_only` set and unset??

Actually it is an API break, because the original code does:

```
    p = params;
    while ((p = OSSL_PARAM_locate(p, OSSL_KDF_PARAM_MODE)) != NULL) {
        if (p->data_type == OSSL_PARAM_UTF8_STRING) {
        ....
        p++;
    }
```
I.e., it cycles through all OSSL_KDF_PARAM_MODE in the array.

Comment (paulidale) [reply test/evp_kdf_test.c:309]: That code isn't in 3.5 that I can see, so it's not been released.  It was added by #27365 which was master only.
So still not an API break.

Nonetheless, it's easy to address and handle multiple copies of the param.

Comment (paulidale) [reply test/evp_kdf_test.c:309]: I can do it here or as a follow up PR.  I'd prefer the latter in order to expedite this and because there are multiple other cases (not KDFs) where we advertise a duplicated param (utf8_string & int) but only ever process one of them.  These would need to be addressed for consistency.

Comment (t8m) [reply test/evp_kdf_test.c:309]: > It was added by #27365 which was master only.
> So still not an API break.

Ahh.... then we are good, I think. I am not even sure it makes sense to support such corner use-case. If anything the caller can always call get params twice.

Comment (paulidale) [reply test/evp_kdf_test.c:309]: A committer or TAC discussion about the sensibility of having it?

It wouldn't be hard to implement but it should be done after some thought and if it is done it must be done everywhere this possibility arises.

Comment (paulidale) [reply test/evp_kdf_test.c:309]: I've raised the decision about which way to process these in #28099.

**Comment by paulidale** [comment]
Still waiting on second approval...

**Comment by paulidale** [comment]
Still waiting...
Please??

**Comment by beldmit** [comment]
@paulidale I see personal problems with this PR. I can't review it, sorry. 2000+ LOC changes, even more or less mechanical, is beyond my concentration window. 

And, last but not least, I'm not a big fan of pre-processing C files. We already have quite complicated macros and this pre-processing adds one more level of indirection. 

Sorry for that

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.
